File: packages/react-native-notificare-monetize/android/src/main/java/re/notifica/monetize/react_native/NotificareMonetizeModule.kt
package re.notifica.monetize.react_native

import android.os.Handler
import android.os.Looper
import androidx.lifecycle.Observer
import com.facebook.react.bridge.*
import re.notifica.Notificare
import re.notifica.NotificareCallback
import re.notifica.internal.NotificareLogger
import re.notifica.monetize.NotificareMonetize
import re.notifica.monetize.ktx.monetize
import re.notifica.monetize.models.NotificareProduct
import re.notifica.monetize.models.NotificarePurchase

public class NotificareMonetizeModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext), NotificareMonetize.Listener {

    private val productsObserver = Observer<List<NotificareProduct>> { products ->
        if (products == null) return@Observer

        try {
            val arguments = Arguments.createArray()
            products.forEach { arguments.pushMap(it.toJson().toReactMap()) }

            EventBroker.dispatchEvent("products_updated", arguments)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the products_updated event.", e)
        }
    }

    private val purchasesObserver = Observer<List<NotificarePurchase>> { purchases ->
        if (purchases == null) return@Observer

        try {
            val arguments = Arguments.createArray()
            purchases.forEach { arguments.pushMap(it.toJson().toReactMap()) }

            EventBroker.dispatchEvent("purchases_updated", arguments)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchases_updated event.", e)
        }
    }

    override fun getName(): String = "NotificareMonetizeModule"

    override fun initialize() {
        super.initialize()

        EventBroker.setup(reactApplicationContext)

        onMainThread {
            Notificare.monetize().observableProducts.observeForever(productsObserver)
            Notificare.monetize().observablePurchases.observeForever(purchasesObserver)
        }
    }

    override fun invalidate() {
        super.invalidate()

        onMainThread {
            Notificare.monetize().observableProducts.removeObserver(productsObserver)
            Notificare.monetize().observablePurchases.removeObserver(purchasesObserver)
        }
    }

    @ReactMethod
    public fun addListener(@Suppress("UNUSED_PARAMETER") eventName: String) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    public fun removeListeners(@Suppress("UNUSED_PARAMETER") count: Int) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    // region Notificare Monetize

    @ReactMethod
    public fun getProducts(promise: Promise) {
        try {
            val products = Notificare.monetize().products

            val arguments = Arguments.createArray()
            products.forEach { arguments.pushMap(it.toJson().toReactMap()) }

            promise.resolve(arguments)
        } catch (e: Exception) {
            promise.reject(DEFAULT_ERROR_CODE, e)
        }
    }

    @ReactMethod
    public fun getPurchases(promise: Promise) {
        try {
            val purchases = Notificare.monetize().purchases

            val arguments = Arguments.createArray()
            purchases.forEach { arguments.pushMap(it.toJson().toReactMap()) }

            promise.resolve(arguments)
        } catch (e: Exception) {
            promise.reject(DEFAULT_ERROR_CODE, e)
        }
    }

    @ReactMethod
    public fun refresh(promise: Promise) {
        Notificare.monetize().refresh(object : NotificareCallback<Unit> {
            override fun onSuccess(result: Unit) {
                promise.resolve(null)
            }

            override fun onFailure(e: Exception) {
                promise.reject(DEFAULT_ERROR_CODE, e)
            }
        })
    }

    @ReactMethod
    public fun startPurchaseFlow(data: ReadableMap, promise: Promise) {
        val activity = currentActivity ?: run {
            promise.reject(DEFAULT_ERROR_CODE, "Cannot start a purchase without an activity.")
            return
        }

        val product: NotificareProduct = try {
            NotificareProduct.fromJson(data.toJson())
        } catch (e: Exception) {
            promise.reject(DEFAULT_ERROR_CODE, e)
            return
        }

        Notificare.monetize().startPurchaseFlow(activity, product)
    }

    // endregion

    // region NotificareMonetize.Listener

    override fun onBillingSetupFinished() {
        try {
            EventBroker.dispatchEvent("billing_setup_finished", null)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the billing_setup_failed event.", e)
        }
    }

    override fun onBillingSetupFailed(code: Int, message: String) {
        try {
            val arguments = Arguments.createMap().apply {
                putInt("code", code)
                putString("message", message)
            }

            EventBroker.dispatchEvent("billing_setup_failed", arguments)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the billing_setup_failed event.", e)
        }
    }

    override fun onPurchaseFinished(purchase: NotificarePurchase) {
        try {
            EventBroker.dispatchEvent("purchase_finished", purchase.toJson().toReactMap())
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_finished event.", e)
        }
    }

    override fun onPurchaseRestored(purchase: NotificarePurchase) {
        try {
            EventBroker.dispatchEvent("purchase_restored", purchase.toJson().toReactMap())
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_restored event.", e)
        }
    }

    override fun onPurchaseCanceled() {
        try {
            EventBroker.dispatchEvent("purchase_canceled", null)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_canceled event.", e)
        }
    }

    override fun onPurchaseFailed(code: Int, message: String) {
        try {
            val arguments = Arguments.createMap().apply {
                putInt("code", code)
                putString("message", message)
            }

            EventBroker.dispatchEvent("purchase_failed", arguments)
        } catch (e: Exception) {
            NotificareLogger.error("Failed to emit the purchase_failed event.", e)
        }
    }

    // endregion

    public companion object {
        internal const val DEFAULT_ERROR_CODE = "notificare_error"

        internal fun onMainThread(action: () -> Unit) = Handler(Looper.getMainLooper()).post(action)
    }
}

File: packages/react-native-notificare-monetize/ios/NotificareMonetizeModule.swift
import React
import NotificareKit
import NotificareMonetizeKit

private let DEFAULT_ERROR_CODE = "notificare_error"

@objc(NotificareMonetizeModule)
class NotificareMonetizeModule: RCTEventEmitter {

    private var hasListeners = false
    private var eventQueue = [(name: String, payload: Any?)]()

    override init() {
        super.init()

        Notificare.shared.monetize().delegate = self
    }

    override class func requiresMainQueueSetup() -> Bool {
        return true
    }

    override func startObserving() {
        hasListeners = true

        if !eventQueue.isEmpty {
            NotificareLogger.debug("Processing event queue with \(eventQueue.count) items.")
            eventQueue.forEach { sendEvent(withName: $0.name, body: $0.payload)}
            eventQueue.removeAll()
        }
    }

    override func stopObserving() {
        hasListeners = false
    }

    override func supportedEvents() -> [String] {
        return [
            "re.notifica.monetize.billing_setup_finished",
            "re.notifica.monetize.billing_setup_failed",
            "re.notifica.monetize.products_updated",
            "re.notifica.monetize.purchases_updated",
            "re.notifica.monetize.purchase_finished",
            "re.notifica.monetize.purchase_restored",
            "re.notifica.monetize.purchase_canceled",
            "re.notifica.monetize.purchase_failed",
        ]
    }

    private func dispatchEvent(_ name: String, payload: Any?) {
        if hasListeners {
            sendEvent(withName: name, body: payload)
        } else {
            eventQueue.append((name: name, payload: payload))
        }
    }

    // MARK: - Notificare Monetize

    @objc func getProducts(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        do {
            let products = try Notificare.shared.monetize().products.map { try $0.toJson() }
            resolve(products)
        } catch {
            reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
        }
    }
    
    @objc func getPurchases(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        do {
            let purchases = try Notificare.shared.monetize().purchases.map { try $0.toJson() }
            resolve(purchases)
        } catch {
            reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
        }
    }
    
    @objc func refresh(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) -> Void {
        Notificare.shared.monetize().refresh { result in
            switch result {
            case .success:
                resolve(nil)
            case let .failure(error):
                reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            }
        }
    }
    
    @objc func startPurchaseFlow(_ data: [String: Any], resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        let product: NotificareProduct
        
        do {
            product = try NotificareProduct.fromJson(json: data)
        } catch {
            reject(DEFAULT_ERROR_CODE, error.localizedDescription, nil)
            return
        }
        
        Notificare.shared.monetize().startPurchaseFlow(for: product)
    }
}

extension NotificareMonetizeModule: NotificareMonetizeDelegate {
    func notificare(_ notificareMonetize: NotificareMonetize, didUpdateProducts products: [NotificareProduct]) {
        do {
            dispatchEvent("re.notifica.monetize.products_updated", payload: try products.map { try $0.toJson() })
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.monetize.products_updated event.", error: error)
        }
    }
    
    func notificare(_ notificareMonetize: NotificareMonetize, didUpdatePurchases purchases: [NotificarePurchase]) {
        do {
            dispatchEvent("re.notifica.monetize.purchases_updated", payload: try purchases.map { try $0.toJson() })
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.monetize.purchases_updated event.", error: error)
        }
    }
    
    func notificare(_ notificareMonetize: NotificareMonetize, didFinishPurchase purchase: NotificarePurchase) {
        do {
            dispatchEvent("re.notifica.monetize.purchase_finished", payload: try purchase.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.monetize.purchase_finished event.", error: error)
        }
    }
    
    func notificare(_ notificareMonetize: NotificareMonetize, didRestorePurchase purchase: NotificarePurchase) {
        do {
            dispatchEvent("re.notifica.monetize.purchase_restored", payload: try purchase.toJson())
        } catch {
            NotificareLogger.error("Failed to emit the re.notifica.monetize.purchase_restored event.", error: error)
        }
    }
    
    func notificareDidCancelPurchase(_ notificareMonetize: NotificareMonetize) {
        dispatchEvent("re.notifica.monetize.purchase_canceled", payload: nil)
    }
    
    func notificare(_ notificareMonetize: NotificareMonetize, didFailToPurchase error: Error) {
        dispatchEvent("re.notifica.monetize.purchase_failed", payload: ["errorMessage": error.localizedDescription])
    }
}
